pyramid_zipkin/tween.py

```diff
@@ -198,7 +198,8 @@ def tween(request: Request) -> Response:
             except Exception as e:
                 zipkin_context.update_binary_annotations({
                     'error.type': type(e).__name__,
-                    'response_status_code': '500'
+                    'response_status_code': '500',
+                    'http.response.status_code': '500',
```

Comment:
while we're here, lets add `error.type` from the [conventions](https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-server-semantic-conventions) too. I wonder if we should even try and add the exception data into the span? otel states that we should be adding exceptions as an event on the span - otel seems to turn zipkin annotations into events, so lets add that? no problem if you think we should add it in a separate PR - just thinking it'd be good to make the most of a new release (there are probably other binary annotations it'd be good to add too)

Reply:
I think adding Events when an [Exception](https://opentelemetry.io/docs/specs/otel/trace/exceptions/) occurs will be good, this will show as `Logs` in an OTel compliant system like Jaeger.
![Screenshot 2024-06-27 at 18 39 11](https://github.com/Yelp/pyramid_zipkin/assets/36337836/9407be7b-cc3a-475f-b075-43ec42424d36)

In raw OTLP, the screenshot will be represented as in the otlp span

```json
"events": [
    {
        "name": "exception-event",
        "timeUnixNano": "1694705188862054000",
        "attributes": [
            {
                "key": "exception.type",
                "value": {
                    "stringValue": "java.lang.IllegalArgumentException"
                }
            },
            {
                "key": "exception.stacktrace",
                "value": {
                    "stringValue": "java.lang.IllegalArgumentException: Invalid claim id\n\tat biz.claims.ClaimService.get(ClaimService.java:42)..."
                }
            }
        ]
    }
],
```
Unfortunately, Zipkin does not have the concept of [Events](https://opentelemetry.io/docs/specs/otel/trace/sdk_exporters/zipkin/#events) and the best we can do is to add the exceptions as zipkin tags or attributes for now, the main difference is that the tag will not have a timestamp of when that event occurred

![Screenshot 2024-06-27 at 18 48 12](https://github.com/Yelp/pyramid_zipkin/assets/36337836/438b6e6a-0f37-4cfa-ad38-b0e5a5a7bf3e)

It might be useful to update the PR to make as many tags compliant with OTels Stable semantic conventions for HTTP server spans as well. https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-server 
Let me know what you think

Reply:
> Unfortunately, Zipkin does not have the concept of [Events](https://opentelemetry.io/docs/specs/otel/trace/sdk_exporters/zipkin/#events) and the best we can do is to add the exceptions as zipkin tags or attributes for now, the main difference is that the tag will not have a timestamp of when that event occurred

reading the link, it mentions that otel represents events as zipkin annotations https://zipkin.io/pages
- we do this internally already for `py_zipkin.logging_end` which show up in the same logs section of the UI in jaeger

Reply:
> It might be useful to update the PR to make as many tags compliant with OTels Stable semantic conventions for HTTP server spans as well. https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#http-server

this would be fantastic if you're willing!

Reply:
Unfortunately non otel spec compliant tag names such as client_status_code, response_status_code might be present in other parts of the code. And there might be some dependencies on them (code, dashboards). 
I have not looked at these closely but some examples:
zipkin-proxy: https://sourcegraph.yelpcorp.com/packages/zipkin-proxy@3d56a45dca4160359eb91a9141154dd5e406e493/-/blob/src/metrics/service.go?L159:30-159:50
syslog2scribe: https://sourcegraph.yelpcorp.com/misc/syslog2scribe@413bf9fc29bc52a1bebe93ef55b267deb07cfab3/-/blob/src/util/service_syslog_zipkin.cpp?L45:21-45:41

Reply:
> reading the link, it mentions that otel represents events as zipkin annotations https://zipkin.io/pages. we do this internally already for `py_zipkin.logging_end` which show up in the same logs section of the UI in jaeger

Yes, zipkin sort of has OTel events that show up as `Logs` but ONLY a single `event` value with the timestamp. For example, raw zipkin annotations will be represented as below
```json
"annotations": [
      {
        "timestamp": 1719513676366000,
        "value": "exception occured"
      }
    ]
```

While OTel represents a single timestamped event with annotations (dict of key-value pairs).
The best we can do with zipkin is maybe to join the `exception.type` and `exception.stacktrace` into just one event like this

```json
"annotations": [
    {
        "timestamp": 1697040000000000,
        "value": "exception.type: java.lang.IllegalArgumentException\nexception.stacktrace: java.lang.IllegalArgumentException: Invalid claim id\n\tat biz.claims.ClaimService.get(ClaimService.java:42)\n\tat biz.claims.ClaimService.get(ClaimService.java:36)\n\tat biz.claims.ClaimService.get(ClaimService.java:30)\n\tat biz.claims.ClaimService.get(ClaimService.java:24)\n\tat biz.claims.ClaimService.get(ClaimService.java:18)\n\tat biz.claims.ClaimService.get(ClaimService.java:12)\n\tat biz.claims.ClaimService.get(ClaimService.java:6)\n\tat biz.claims.ClaimService.get(ClaimService.java:3)\n\tat biz.claims.ClaimService.get(ClaimService.java:1"
    }
]
```

Reply:
> Unfortunately non otel spec compliant tag names such as client_status_code, response_status_code might be present in other parts of the code. And there might be some dependencies on them (code, dashboards). 
> 
> I have not looked at these closely but some examples:
> 
> zipkin-proxy: https://sourcegraph.yelpcorp.com/packages/zipkin-proxy@3d56a45dca4160359eb91a9141154dd5e406e493/-/blob/src/metrics/service.go?L159:30-159:50
> 
> syslog2scribe: https://sourcegraph.yelpcorp.com/misc/syslog2scribe@413bf9fc29bc52a1bebe93ef55b267deb07cfab3/-/blob/src/util/service_syslog_zipkin.cpp?L45:21-45:41
> 
> 

Plan is to add the otel specific tags as an addition to the existing tags that have references in other parts of the code like response_status_code. At least until full migration

Reply:
alternatively, can we log multiple annotations with the same timestamp?

Reply:
No, not with zipkin since it only allows a single [string value](https://zipkin.io/zipkin-api/#/): 
![Screenshot 2024-06-28 at 11 23 08](https://github.com/Yelp/pyramid_zipkin/assets/36337836/59d2b46b-4e18-4e30-b3a0-745f6054ad3a)

Reply:
right, so annotations need unique timestamps. unhelpful. I think your plan of adding the exception stacktrace as a tag is probably the next best option then